From a0e787a233b0afe79c8b28302f5e99a6388ddc87 Mon Sep 17 00:00:00 2001
From: Dmitry Petrochenko <dmitry.petrochenko@intel.com>
Date: Mon, 19 Aug 2013 16:33:45 +0700
Subject: Dalvik: PCG: Fix dvmInstanceofNonTrivial result check in check-cast

BZ: 131118

The dvmInstanceofNonTrivial returns zero in case when class cast
exception should be thrown, but check-cast bytecode implementation
thought that zero return value is ok.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-CG
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: If19f7f8dfd2a39cab3d8dbb88bfeb84648c3c79c
Orig-MCG-Change-Id: If57ce219bcf97215a0da580f390ba5aa06a5bdc6
Signed-off-by: Dmitry Petrochenko <dmitry.petrochenko@intel.com>
Reviewed-on: http://android.intel.com:8080/126104
Reviewed-by: Maslov, Sergey V <sergey.v.maslov@intel.com>
Reviewed-by: Beyler, Jean Christophe <jean.christophe.beyler@intel.com>
Reviewed-by: Lupusoru, Razvan A <razvan.a.lupusoru@intel.com>
Reviewed-by: lab_aqa <lab_aqa@intel.com>
Reviewed-by: Semukhina, Elena V <elena.v.semukhina@intel.com>
Reviewed-by: Chen, Dong-Yuan <dong-yuan.chen@intel.com>
Tested-by: Chen, Dong-Yuan <dong-yuan.chen@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/codegen/x86/pcg/LowerOther.cpp |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/vm/compiler/codegen/x86/pcg/LowerOther.cpp b/vm/compiler/codegen/x86/pcg/LowerOther.cpp
index 2e783cc..7ae7908 100644
--- a/vm/compiler/codegen/x86/pcg/LowerOther.cpp
+++ b/vm/compiler/codegen/x86/pcg/LowerOther.cpp
@@ -242,7 +242,7 @@ void dvmCompilerPcgCommonCheckCastInstanceOf (CompilationUnitPCG *cUnit, MIR *mi
     }
     else
     {
-        CGCreateNewInst ("cjcc", "rcrbp", callResult, "eq", CGCreateNewInst ("mov", "i", 0), endLabel, 95);
+        CGCreateNewInst ("cjcc", "rcrbp", callResult, "ne", CGCreateNewInst ("mov", "i", 0), endLabel, 95);
         dvmCompilerPcgGenerateRaiseException (cUnit);
     }
 
-- 
1.7.4.1

